### ISSUE TITLE
Provide a general purpose API for extended plugin configuration

### ISSUE THREAD

**Issue by andrewvc** [elastic/logstash#8582]
Currently plugins like the `grok` filter use files for extended configuration options like https://www.elastic.co/guide/en/logstash/current/plugins-filters-grok.html#plugins-filters-grok-patterns_dir to handle some options. These things are hard to configure centrally however. Ideally there'd be a way to centralize these options, perhaps by letting them be satisfied by an arbitrary URL, not just a central file.

**Comment by jordansissel** [comment]
@andrewvc let's get a list of plugins which use files for extra configuration and see if any obvious abstractions come of it.

> letting them be satisfied by an arbitrary URL

The impetus for this is grok patterns being hard to manage when central config is used. I agree with the problem statement. A URL alone is not sufficient because the "obvious" (to me) place to store patterns is Elasticsearch (when central config is enabled), and Elasticsearch does not serve plain text responses outside of the /_cat api.

**Comment by andrewvc** [comment]
@jordansissel good point! I'm +1 on a more thorough investigation of real world plugin usage first.

WRT Elasticsearch, ES does have URLs, we could have some provision for pulling special JSON fields etc., but I agree that's a question to ask later.
